### PR TITLE
Added backupvault resource to Terraform backupdr product (beta)

### PR DESCRIPTION
**Pull request by liyun-huang**
```release-note:new-resource
`google_backup_dr_backup_vault` (beta)
```

### CODE REVIEW

**Comment by zli82016** [comment]
Rerun the tests

**Comment by zli82016** [comment]
@liyun-huang , do you know how to enable the API on the test project? 

The link `https://console.developers.google.com/apis/api/autopush-backupdr.sandbox.googleapis.com/overview?project=ci-test-project-188019` in the debug [log](https://ffded39fce65fcbb6891cfd1899cdf93b3ee0a4451819de8f4b5bb6-apidata.googleusercontent.com/download/storage/v1/b/ci-vcr-logs/o/beta%2Frefs%2Fheads%2Fauto-pr-11317%2Fartifacts%2F992c3ebb-3311-47d1-9e5e-6fd3b2a18952%2Fbuild-log%2Frecording_build%2FTestAccBackupDRBackupVault_basicUpdate_recording_test.log?jk=AYBlUPCQXj79ihedy4Q4XZpMf6n9o-YC3aNZvKjh2RU5QKRK2iUWyFY_24zGpsh6mX8mz5Q4xc_yisZYkc4hzI2iySR0JIUSnrTcaWghPdCY9__m46PZlq4fZGqrwOPYv-lQqdjvrGbrZcbqzXXRwEULqmGYGLruM0t5tSMfSYpdMaowOdtQGC05J2l4tdkNNCfjfMAbTerWgc1-WZX_9enAn72DWwvv85fZabq7GlvWofsvYMnrkx8tzL_OZDAbsrCvX66okrEZdCgjgQew5Qadh6ngFf9aDZfbJSSy9xNXvrAxTBB0AX88bKIYMDDRbQexW1eyH5xgoF2nAVmDnvZkEwFhTGeLRfTkdpEv69fzh0EiqoS1LZ9scF3Nbv98P0lq_DUdLS9hQgE7EiD_k9llwQfQ5jvWfWONR2tUltld_UGBWK4tQv14eiT5EsSy1DijV8XI-2Ie-ixMI-mFQDtV6JdXVn_kUcC2oLGX3Di9AsVAbUVUrbfGQmufbN-AqvfKnzGzQvZ9dJ_vZEUwC-1jGG1LCZfG0bnwDnIcljopRSRGCYoTiaIU-Kd8Vp71WGcKjA1H7ZkwlmxdFuG73oczwPRltsnV0B3A5rK1I30RZ4999Ra17h09fzKaUXtox1QM7JgrPqjSpuqBxEla1q4mWu3ycuiJ0PJiJaixp__rmmvxNll1fLN4A3vKp5xwpwDeCK3ygfo1h2WtJChvImkOzv3yNu5zVdZ1Aps2zjTJfwVzoiI17mzX6GIKqMshFtPUFCZshfp1iDFVlRLwUOCDzOWz6tQpKRo7fooXJ7PiO_jQzt3Q2vwBsu1ymoD6FTVoPIVPMe4XS34Mvygv54u0CtLD9OtP2Om1gw4PYJSxOdNOiFO5tAdsp-LNPbuzq657xZQLfKx-PIgkWiOfAmoFXqWGuypHEk32a5dQvJwgBZwZl67BHWHFosKMvSSEmm0tyIGX2KFic8gCLoK8RFxXqIEPfyFFHk2MDEQFVuIVRmknamP9e-NoPSUyXEY95AxVVW0PobUajZUMKuEnaDJxUeuTbsXfVdkz6nYY-Im3DuwfC52nsa_U_FdqoZkPusFDF9aNzTgsYKCEA87dYjlTDw4Eyi19VV-K22H_4rAOToWQCkU4GVxUckInjrOCGF1n_CTxvpfS2p07ZBWbajTE6HbGpOOFyVr8oFXdod006RGqlHNR1zU4maneXrL2oXdnLu4-DZKGiI5PqdKv3v7vAtWikH8kmvF9IGmSEUfbZl1y2HKFoXi4MN6SdvnjGQmro9sQJS1cCqZPLnuu-rzcRac41hdvGOnNPyK5THVo1CvXwzu5Gfsa8QDeTc3X9Z2_ErmetAE&isca=1) fails to load.

File: mmv1/products/backupdr/product.yaml [hunk 1]
```diff
@@ -13,13 +13,13 @@
...
 name: BackupDR
-display_name: Backup and DR
+display_name: Backup and DR Service
 scopes:
   - https://www.googleapis.com/auth/cloud-platform
 versions:
   - !ruby/object:Api::Product::Version
     name: beta
-    base_url: https://backupdr.googleapis.com/v1/
+    base_url: https://autopush-backupdr.sandbox.googleapis.com/v1/
```
Comment (zli82016) [inline mmv1/products/backupdr/product.yaml:22]: do you know the reason to change the url?

Comment (liyun-huang) [reply mmv1/products/backupdr/product.yaml:22]: My bad, I had autopush url here for testing. Reverted the url back to prod.

**Comment by zli82016** [comment]
Testes need to be fixed.

File: mmv1/products/backupdr/BackupVault.yaml [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      backup_vault_id: 'tf-test-backup-vault-test%{random_suffix}'
+    test_env_vars:
+      project: :PROJECT_NAME
+  - !ruby/object:Provider::Terraform::Examples
+    min_version: beta
+    name: 'backup_dr_backup_vault_full'
+    primary_resource_id: 'backup-vault-test2'
+    vars:
+      backup_vault_id: 'tf-test2-backup-vault-test%{random_suffix}%{random_suffix}'
+      effective_time: "%{calculated_effective_time}"
```
[the remark this answers is not in the document]
Comment (liyun-huang) [reply mmv1/products/backupdr/BackupVault.yaml]: Renamed to effect_time. It's a variable to keep the date to (today + 24hrs) in test file resource_backup_dr_backup_vault_test.go.erb

**Comment by zli82016** [comment]
The reason for `BackupVault is not enabled on this environment prod-us-central1` could be that this new resource is not in GA. I found the same issue in https://b.corp.google.com/issues/335203829

File: mmv1/products/backupdr/BackupVault.yaml [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+      backup_vault_id: 'backup-vault-test'
+    test_env_vars:
+      project: :PROJECT_NAME
+  - !ruby/object:Provider::Terraform::Examples
+    min_version: beta
+    name: 'backup_dr_backup_vault_full'
+    primary_resource_id: 'backup-vault-test2'
+    vars:
+      backup_vault_id: 'backup-vault-test'
+      effective_time: "%{effective_time}"
```
Comment (zli82016) [inline mmv1/products/backupdr/BackupVault.yaml:26]: `%{effective_time}` is not defined anywhere and cannot be found.

File: mmv1/products/backupdr/BackupVault.yaml [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    test_env_vars:
+      project: :PROJECT_NAME
```
Comment (zli82016) [inline mmv1/products/backupdr/BackupVault.yaml:28]: The full example can be deleted, as it is covered in the update test.

**Review by zli82016 — CHANGES_REQUESTED** [review]

File: mmv1/templates/terraform/examples/backup_dr_backup_vault_full.tf.erb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+resource "google_backup_dr_backup_vault" "<%= ctx[:primary_resource_id] %>" {
```
Comment (zli82016) [inline mmv1/templates/terraform/examples/backup_dr_backup_vault_full.tf.erb:1]: This file can be deleted.

**Review by zli82016 — CHANGES_REQUESTED** [review]

**Comment by zli82016** [comment]
More parameters will be added to the delete url.

File: mmv1/products/backupdr/BackupVault.yaml [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+base_url: projects/{{project}}/locations/{{location}}/backupVaults
+create_url: projects/{{project}}/locations/{{location}}/backupVaults?backupVaultId={{backup_vault_id}}
+update_url: projects/{{project}}/locations/{{location}}/backupVaults/{{backup_vault_id}}?force={{force_update}}
+delete_url: projects/{{project}}/locations/{{location}}/backupVaults/{{backup_vault_id}}?force={{force_delete}}&allowMissing={{allow_missing}}
+self_link: projects/{{project}}/locations/{{location}}/backupVaults/{{backup_vault_id}}
+id_format: projects/{{project}}/locations/{{location}}/backupVaults/{{backup_vault_id}}
+import_format:
+  - projects/{{project}}/locations/{{location}}/backupVaults/{{backup_vault_id}}
+name: BackupVault
+description: Container to store and organize immutable backups.
```
Comment (zli82016) [inline mmv1/products/backupdr/BackupVault.yaml:11]: Are the backups immutable?

File: mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+			bar = "baz"
+		}
+		annotations = {
+			annotations1 = "bar"
+			annotations2 = "baz"
+		}
+		force_update = "false"
+		force_delete = "false"
+		allow_missing = "false"
+		}
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb:70]: nitpick: Can you change the indentations? Each indentation should be two white spaces.

File: mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			bar = "baz1"
+		}
+		annotations = {
+			annotations1 = "bar1"
+			annotations2 = "baz1"
+		}
+		force_update = "false"
+		force_delete = "false"
+		allow_missing = "false"
+		}
```
Comment (zli82016) [inline mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb:94]: nitpick: Can you change the indentations? Each indentation should be two white spaces.

**Review by zli82016 — CHANGES_REQUESTED** [review]

File: mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+			foo = "bar1"
+			bar = "baz1"
+		}
+		annotations = {
+			annotations1 = "bar1"
+			annotations2 = "baz1"
+		}
+		force_update = "false"
+		force_delete = "false"
+		allow_missing = "false"
```
[the remark this answers is not in the document]
Comment (zli82016) [reply mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb:93]: Changing the values to true to test if the update works.

File: mmv1/products/backupdr/BackupVault.yaml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+  - !ruby/object:Api::Type::String
+    name: backupVaultId
+    description: "Required. ID of the requesting object. If auto-generating ID server-side,
+      remove this field and backup_vault_id from the method_signature of Create RPC "
+    url_param_only: true
+    required: true
+    immutable: true
+  - !ruby/object:Api::Type::Boolean
+    name: 'force_update'
+    default_value: false
```
[the remark this answers is not in the document]
Comment (liyun-huang) [reply mmv1/products/backupdr/BackupVault.yaml:104]: I think default_value is needed in our case, the default value is fixed in our API.

File: mmv1/products/backupdr/BackupVault.yaml (hunk 6 above)
Comment (liyun-huang) [reply mmv1/products/backupdr/BackupVault.yaml:11]: Backup vault ensures that backups are both immutable & indelible. Reference: https://docs.google.com/presentation/d/1zFDIQh6DHciMbETvmi0ynt2WXtS4ujGL7y0fsOHnWxs/edit?resourcekey=0-w48LoHQeVwRZJrRL9WbN-Q#slide=id.g2761e27a707_0_558

File: mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb (hunk 9 above)
[the remark this answers is not in the document]
Comment (liyun-huang) [reply mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb:93]: Done

File: mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb (hunk 8 above)
Comment (liyun-huang) [reply mmv1/third_party/terraform/services/backupdr/resource_backup_dr_backup_vault_test.go.erb:94]: Done

**Review by zli82016 — APPROVED** [review]
LGTM. Thanks for adding the new resource.

**Pull request merged**